src/legacy/ui/public/new_platform/new_platform.ts

```diff
@@ -52,3 +73,49 @@ export function getNewPlatform() {
 
   return runtimeContext;
 }
+
+type SetupCallback<T> = (startContext: typeof runtimeContext['setup']) => T;
+type StartCallback<T> = (startContext: typeof runtimeContext['start']) => T;
+
+const onSetupCallbacks: Array<SetupCallback<Promise<unknown>>> = [];
+const onStartCallbacks: Array<StartCallback<Promise<unknown>>> = [];
+
+/**
+ * Register a callback to be called once the new platform is in the
+ * `setup` lifecycle event. Resolves to the return value of the callback.
+ */
+export async function onSetup<T>(callback: SetupCallback<T>): Promise<T> {
+  if (runtimeContext.setup.core !== null) {
+    return callback(runtimeContext.setup);
+  }
+
+  return new Promise((resolve, reject) => {
+    onSetupCallbacks.push(async (setupContext: typeof runtimeContext['setup']) => {
+      try {
```

Comment:
optional nit: promise based api is less verbose
``` js
 onSetupCallbacks.push((setupContext: typeof runtimeContext['setup']) => callback(setupContext).then(resolve, reject)
```

Reply:
I don't think this is possible since `callback` does not necessarily return a Promise.

Reply:
ok, I see
```typescript
type SetupCallback<T> = (startContext: typeof runtimeContext['setup']) => T;
```